### ISSUE TITLE
[Security] Input state polluted.

### ISSUE THREAD

**Issue by PhilETaylor** [joomla/joomla-cms#37180]
Edit: initially this issue was opened to discuss a historic security issue that allowed login using a get request that it appeared had not been fixed. I was wrong, at the time it was fixed, and I apologise for my mistake.

During discussion below, it was however discovered, that this is a new issue regarding a polluted input state, a more serious issue that effects the whole of Joomla and not just login feature! and not as originally thought, just a failure to fix the previously reported issue.

Issue described here https://github.com/joomla/joomla-cms/issues/37180#issuecomment-1057821611

**Comment by brianteeman** [comment]
but you need to know the username, the password and the token ?

**Comment by PhilETaylor** [comment]
You miss the point.

Name me 10 web sites where you can login with a GET request? Your bank? PayPal? Amazon? Apple? You can't. 

Because it's stupid, wrong, and goes against every single security practice. 

Also all query data is logged into the logs....

**Comment by brianteeman** [comment]
it was a question

**Comment by PhilETaylor** [comment]
Then let me answer the question - yes, you require the correct username, password and correct and valid token in order to achieve authentication and receive a new authenticated session cookie 

But, I could run a million GET requests a second to guess the credentials... 

You will be interested in my next Issue ignored by the JSST, where I DONT need your password, and I can still login as you... if a single 0 is changed to a 1 in the db .... they keep telling me its going to be fixed but we are coming up on 4 years now...

**Comment by kitepascal** [comment]
How could it be fixed? I would donate for that because J!4 is going to be my strongest power horse.

**Comment by PhilETaylor** [comment]
Probably one line of code somewhere near https://github.com/joomla/joomla-cms/blob/8909b0b9420ef35ffacaa3f2224e994d4aea7687/administrator/components/com_login/src/Model/LoginModel.php#L39 needs to change... but hey...

**Comment by kitepascal** [comment]
Perfect, may you start a Pull request?

I understand your frustratiom - sometimes it's not easy to work in a team.

**Comment by PhilETaylor** [comment]
> Perfect, may you start a Pull request?

I reported it to the official Joomla Security Strike Team to resolve on 9th June 2020 - they have had years to resolve it in a private pull request.

**Comment by kitepascal** [comment]
Additional lines of code for security are always good imho.

**Comment by SniperSister** [comment]
I'm unable to replicate the issue on a current 4.x installation, I get a "The security token did not match. The request was aborted to prevent any security breach. Please try again." message.

A PR fixing the issue was #31236 where you @PhilETaylor confirmed that the issue has been fixed. Do I miss anything here?

**Comment by PhilETaylor** [comment]
I have just confirmed this again, on a Joomla 4.1.0 Stable version on a totally different computer from yesterday. 

**There is still an issue** to be resolved it seems. 

Im sorry that the head of the Joomla Security Strike Team is more interested in spreading lies and malicious slander in German Language Chats thinking that I will not understand what is being said, instead of actually investigating and fixing security issues. 

as stated in https://github.com/joomla/joomla-cms/pull/31236#issuecomment-716176832 - "hint, there is another security issue"

In that post I also revealed other things that the JSST should be taking into account - and was again ignored. 


> Did you take a wider look, or specific to the report? What about the other 13 uses of checkToken('request') - have you ensured those are also not introducing a security issue? (hint, there is another security issue

> Should a GET have the ability to set the default "home" property for a list of items by GET request, when it should only be able to be set by POST (@since 1.6)

> Surely the ones using 'request' should be more specifically a "get" or a "post" (which is the default when no method set)? a method that can be called by GET and POST is probably badly coded I guess.

> $this->checkToken('request'); is also used in the "Method to confirm the password request." - is that ok in context?

**Comment by kitepascal** [comment]
I reproduced that and also think that login via GET /administrator/?option=com_login&task=login&username=adminuser&passwd=password&a148c641d9ce8e27db0f840b07782522=1

should not be possible.
You get the needed token by accessing /administrator
And then you can fire your get request(s).

Why or for what is that needed?

I also followed the Reference Links above where it is clearly described as bad practice.

**Comment by SniperSister** [comment]
> There is still an issue to be resolved it seems.

Interesting. I was unable to reproduce the issue on multiple installations, but can indeed confirm that it still occurs on one machine. I'm already on it, investigating the actual root cause.

So: thanks for bringing that up again!

**Comment by PhilETaylor** [comment]
>  investigating the actual root cause.

Pollution of the input state where GET variables are being incorrectly stored as POST variables as well, deep within Joomla Framework components as well as the over use/incorrect use of the `getInputForRequestMethod` within com_login when only ever POST request methods should be taken into account for handling credentials. 

For example, when on url http://127.0.0.1/administrator/index.php?option=com_login&task=login&username=admin&passwd=adminadminadmin&156648435e27144fd76326f019d9ab33=1

the Input object looks like this. 

<img width="598" alt="Screen Shot 2022-03-03 at 08 59 00" src="https://user-images.githubusercontent.com/400092/156531243-f6701627-9e89-4c56-8072-2ecb3ec0dd87.png">

**Comment by PhilETaylor** [comment]
> A PR fixing the issue was https://github.com/joomla/joomla-cms/pull/31236 where you @PhilETaylor confirmed that the issue has been fixed. Do I miss anything here?

Yes what you are missing is that that DID FIX THE ISSUE reported AT THE TIME, this is why I stated AT THE TIME it was fixed - but this now seems to be a NEW ISSUE with the input state being corrupt. 

This can be proved with the above image, the change you made previously was to check the token in the POST instead of a token in the "request" (meaning either GET or POST).  The token is STILL BEING CHECKED in the POST of the Input, but because something is corrupting the state of the Input object now, that token, sent by the url as a GET, is being stored in the post.

This opens up a whole new can of worms... 

Basically right now, in Joomla 4.1, it seems, that you can bypass POST token checks, by passing the token in the url (GET) requests. I can think of other serious security issues this opens up. Doh. 

I wonder what other knock on effects a polluted Input state could cause - I'll think on that in the shower and over coffee - but I think today could become a very bad day for Joomla 4....

**Comment by PhilETaylor** [comment]
> This opens up a whole new can of worms...

> I think today could become a very bad day for Joomla 4....

Confirmed.

Lets create a new user with a GET request...  (Site configured to allow User Registrations, which is disabled out of the box) 

```
http://127.0.0.1/index.php/component/users/?task=registration.register&Itemid=101&66214f147fbd3c99b5494bb73e0a5e51=1&jform[name]=TESTUSER&jform[username]=PASSWORD&jform[password1]=12345678901234567890&jform[password2]=12345678901234567890&jform[email1]=nonsuch@example.com
```

<img width="1288" alt="Screen Shot 2022-03-03 at 09 12 39" src="https://user-images.githubusercontent.com/400092/156533512-b9e1f2f3-46ad-4a23-aa6a-179fbbf264c7.png">

**Comment by PhilETaylor** [comment]
This was initially broken 6 years ago (Sep 17, 2016 ) by @mbabker (who I have spoken to today, so no shame dropping here) in commit in the joomla-framework/input repo 2.0-dev branch 

https://github.com/joomla-framework/input/commit/9135fe4efe961ef1573b0e5765740cbbbf4d6f37

specifically the change:
 
<img width="1053" alt="Screen Shot 2022-03-03 at 12 59 06" src="https://user-images.githubusercontent.com/400092/156569392-2d3396ec-e7f6-4c8d-9bb8-1fe3605742e2.png">

More specifically the change to be:
```php
$this->data    = empty($source) ? $_REQUEST : $source;
```

at the same time as changing the constructor to be 

`public function __construct(array $source = [], array $options = [])`

HOWEVER - then 4 years ago (May 26, 2018) on merging master into 2.0-dev he overwrote the constructor in https://github.com/joomla-framework/input/commit/1ed80e8a119ab26a1c366e470e967786758e2b68 so it was reverted to 

```php
public function __construct($source = null, array $options = [])
```

thus introducing a bug as the constructor and the empty check were not doing the same thing. 

Jun 9, 2020, 2 years ago @wilsonge moved Joomla 4 to using the Joomla Framework ~2.0@beta https://github.com/joomla/joomla-cms/commit/d8109cb3cb3d06943bc8bbcaac4236cc3761016c - this first introduced the bug to Joomla 4 codebase. 

Now, on Aug 16, 2021 @wilsonge merged the Joomla Framework 2.0 Stable has cementing this bug into Joomla 4. 

The fix therefore, is to revisit the constructors of all the classes in the Joomla-framework/input package that were wrongly changed in https://github.com/joomla-framework/input/commit/1ed80e8a119ab26a1c366e470e967786758e2b68 and revert that change, ensuring that any further call to the `$source` (which is an array, the contents of $_POST for example) are treated as a `null|array` and not treated with `empty()` which will return TRUE for `empty([])` (which is the root bug here, if the POST is empty when you try to call `Factory::$application->input->post->get('somethin')` then it will use the `$_REQUEST` global, which has the `GET` params in it!!! 

Case Solved.

**Comment by PhilETaylor** [comment]
Tagging @nibra here.

**Comment by PhilETaylor** [comment]
Also this line in Joomla needs to ensure that it only ever uses the POST data and never GET data

https://github.com/joomla/joomla-cms/blob/4006e418db836cdcc0dba78a45d846a9f667b2b0/administrator/components/com_login/src/Model/LoginModel.php#L39

**Comment by PhilETaylor** [comment]
This should also be a complete release blocker!

**Comment by PhilETaylor** [comment]
Monkey see Monkey Do. https://github.com/joomla-framework/input/pull/36

**Comment by SniperSister** [comment]
Full transparency: The underlying issue was originally discovered by Nic who reported the issue in Nov 2021, see
https://github.com/joomla/joomla-cms/issues/35541#issuecomment-961934721

I failed to properly recognize the effects of that issue at the time of the report and added it to the "needs closer check"-list.

**Comment by PhilETaylor** [comment]
So the official Joomla Security Strike Team have been aware of a security issue since November 2021 - did not correctly triage it - and did nothing about it but put it on a list… got it…

**Comment by PhilETaylor** [comment]
oh would you look at that... I was reporting this issue without even knowing it October 2020... https://github.com/joomla/joomla-cms/issues/30971 #facepalm me.
